### PR TITLE
Feature: Download / Upload storage files in batch

### LINKED ISSUE #998: [Storage] Support downloading multiple blobs/files with a single call
Since blob names allow slashes to simulate directories for organizational purposes, it would be nice if we could download multiple blobs by specifying a prefix (ex: CreateVM_2016-07-01*). This would require the downloaded file name to default as described in issue #995.

For files, you would download multiple files not by specifying a file prefix (although you could), but by specifying the directory path.

### PR DESCRIPTION
**Pull request by troydai**
Fixes #998.

### CODE REVIEW

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_params.py [hunk 1]
```diff
@@ -278,6 +279,37 @@ def register_source_uri_arguments(scope):
...
+
+# FILE DOWNLOAD-BATCH PARAMETERS
+register_extra_cli_argument('storage file download-batch', 'source',
+                            options_list=('--source', '-s'),
+                            required=True,
+                            validator=process_file_download_batch_parameters)
+register_cli_argument('storage file download-batch', 'destination',
+                      options_list=('--destination', '-d'),
+                      required=True)
+register_cli_argument('storage file download-batch', 'source_share_name', ignore_type)
```
Comment (tjprescott) [inline src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_params.py]: Same here--I'd argue for handling this in the custom method signature.

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py [hunk 2]
```diff
@@ -408,6 +392,64 @@ def process_blob_copy_batch_namespace(namespace):
     if namespace.prefix is None and not namespace.recursive:
         raise ValueError('incorrect usage: --recursive | --pattern PATTERN')
 
+
+def process_file_upload_batch_parameters(namespace):
+    """Process the parameters of storage file batch upload command"""
+
+    # 1. quick check
+    if not os.path.exists(namespace.source):
+        raise ValueError('incorrect usage: source {} does not exist'.format(namespace.source))
```
Comment (tjprescott) [inline src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py]: In the blob method above you use:
`(incorrect usage: --destination must be an existing directory')`

This is inconsistent.

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py [hunk 3]
```diff
@@ -408,6 +392,64 @@ def process_blob_copy_batch_namespace(namespace):
...
+
+    if not os.path.isdir(namespace.source):
+        raise ValueError('incorrect usage: source must be a directory')
```
Comment (tjprescott) [inline src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py]: Also inconsistent: "incorrect usage: --source must be a directory" (or change the message in the blob method works too)

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py [hunk 4]
```diff
@@ -408,6 +392,64 @@ def process_blob_copy_batch_namespace(namespace):
...
+
+    # 2. try to extract account name and container name from destination string
+    from .storage_url_helpers import StorageResourceIdentifier
+    identifier = StorageResourceIdentifier(namespace.destination)
+    if not identifier.is_url():
+        namespace.destination_share_name = namespace.destination
```
Comment (tjprescott) [inline src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py]: With this, it really doesn't make sense to register destination as extra, ignore destination_share_name and then simply transfer the value from one to the other. I would simplify this.

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py [hunk 5]
```diff
@@ -408,6 +392,64 @@ def process_blob_copy_batch_namespace(namespace):
...
+        namespace.account_name = identifier.account_name
+
+    # 3. collect the files to upload
+    namespace.source = os.path.realpath(namespace.source)
+    namespace.source_files = [c for c in glob_files_locally(namespace.source, namespace.pattern)]
+
+    # 4. clean up
+    del namespace.destination
+    del namespace.source
+    del namespace.pattern
```
Comment (tjprescott) [inline src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py]: No need to delete source and destination if you refactor.

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py [hunk 6]
```diff
@@ -408,6 +392,64 @@ def process_blob_copy_batch_namespace(namespace):
...
+    if not os.path.isdir(namespace.source):
+        raise ValueError('incorrect usage: source must be a directory')
+
+    # 2. try to extract account name and container name from destination string
+    from .storage_url_helpers import StorageResourceIdentifier
+    identifier = StorageResourceIdentifier(namespace.destination)
+    if not identifier.is_url():
+        namespace.destination_share_name = namespace.destination
+    elif identifier.filename or identifier.directory:
+        raise ValueError('incorrect usage: destination must be a file share url')
```
Comment (tjprescott) [inline src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py]: Not consistent with the message @ L434

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py (hunk 2 above)
Comment (troydai) [inline src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py]: update the usage message in blob instead.

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py (hunk 3 above)
Comment (troydai) [inline src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py]: update the usage message in blob instead.

**Comment by troydai** [comment]
@tjprescott updated.

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_help.py [hunk 7]
```diff
@@ -181,20 +181,20 @@
...
-          short-summary: The directory where the files to be uploaded.
+          short-summary: The directory from which the files should to be uploaded.
         - name: --destination -d
           type: string
-          short-summary: The string represents the destination of this upload operation. The source
-                         can be the container URL or the container name. When the source is the
-                         container URL, the storage account name will parsed from the URL.
+          short-summary: The string represents the destination of this upload operation. The
+                         destination can be the file share URL or the share name. When the source
+                         is the share URL, the storage account name will parsed from the URL.
```
Comment (tjprescott) [inline src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_help.py]: Does the "file share URL" also account for directory path? My URL could be "https://file.account.azure.com/myshare/dir1/dir2/dir3". That was supported with the stand-alone upload. In this case the "path" parameter in the upload call would be "dir1/dir2/dir3".

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_help.py [hunk 8]
```diff
@@ -213,9 +213,9 @@
     parameters:
         - name: --source -s
           type: string
-          short-summary: The string represents the destination of this upload operation. The source
-                         can be the container URL or the container name. When the source is the
-                         container URL, the storage account name will parsed from the URL.
+          short-summary: The string represents the destination of this file download operation. The
```
Comment (tjprescott) [inline src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_help.py]: > The string represents the destination of this file download operation

...except that it doesn't. :wink: This is the source parameter.

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py [hunk 9]
```diff
@@ -406,48 +406,36 @@ def process_file_upload_batch_parameters(namespace):
...
     from .storage_url_helpers import StorageResourceIdentifier
     identifier = StorageResourceIdentifier(namespace.destination)
-    if not identifier.is_url():
-        namespace.destination_share_name = namespace.destination
-    elif identifier.filename or identifier.directory:
-        raise ValueError('incorrect usage: destination must be a file share url')
-    else:
-        namespace.destination_share_name = identifier.share
+    if identifier.is_url():
+        if identifier.filename or identifier.directory:
```
Comment (tjprescott) [inline src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py]: Actually, this *should* work if identifier is a directory. If you create a directory in your share called "uploaded" and you want everything to be uploaded there, you should be able to. Forcing the user to batch-upload only to a share is like saying you can only put files on your machine at the root directory... but on any drive you want!

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py [hunk 10]
```diff
@@ -406,48 +406,36 @@ def process_file_upload_batch_parameters(namespace):
...
+            raise ValueError('incorrect usage: destination must be a file share url')
 
-    if not namespace.account_name:
-        namespace.account_name = identifier.account_name
+        namespace.destination = identifier.share
```
Comment (tjprescott) [inline src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py]: Yes, I think you also need to parse any directory subpath from the URL and apply that as well.

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py [hunk 11]
```diff
@@ -406,48 +406,36 @@ def process_file_upload_batch_parameters(namespace):
...
     identifier = StorageResourceIdentifier(namespace.source)
-    if not identifier.is_url():
-        namespace.source_share_name = namespace.source
-    elif identifier.filename or identifier.directory:
-        raise ValueError('incorrect usage: --source should be either share URL or name')
-    else:
-        namespace.source_share_name = identifier.share
+    if identifier.is_url():
+        if identifier.filename or identifier.directory:
+            raise ValueError('incorrect usage: source should be either share URL or name')
```
Comment (tjprescott) [inline src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py]: Similar logic as above. If I don't want to recursively download the entire share, but only a directory within a share, I should be able to do that.

Comment (troydai) [inline src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py]: use can use pattern to do that.
```
--pattern folder/*
```

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py (hunk 10 above)
Comment (troydai) [inline src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py]: As you have pointed out in other comments, in this implementation we accept share only. If user want to download specific directory in a share he can do that with the pattern.

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_help.py (hunk 7 above)
Comment (troydai) [inline src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_help.py]: No. We accept URL to a share only. It is not supported to specify directory in URL yet. We can expand this feature.

File: src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py (hunk 10 above)
Comment (tjprescott) [inline src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py]: Okay, wanted to make sure the functionality is still present. I still think everything (account name, share, directory path, and pattern) should be able to be parsed from the URL, but as long as the functionality is present we can check this in.

**Comment by tjprescott** [comment]
One of the help entries still needs to be updated, but it's clear that the ability to specify a directory is still present, so I'm okay with merging this and tweaking it later.

**Pull request merged**